packages/compartment-mapper/src/zip/buffer-reader.js

```diff
@@ -0,0 +1,225 @@
+// @ts-check
+/* eslint no-bitwise: ["off"] */
+
+const q = JSON.stringify;
+
+// This is a mask of all the bits of the high 32 bits of a 64 bit integer that
+// indicate a number that cannot be stored in the 53 bits of integer precision
+// in a double precision floating point number.
+const bitsTooHigh = (-1 << (53 - 32)) >> 0;
+
+export class BufferReader {
+  /**
+   * @param {ArrayBuffer} data
+   */
+  constructor(data) {
+    this.data = new Uint8Array(data);
+    this.length = this.data.length;
+    this.index = 0;
+    this.offset = 0;
```

Comment:
I see from other callers that `this.offset` is meant to compensate for e.g. prepended non-ZIP data at the beginning of the file. I see no method to set the offset, so I'm guessing callers are expected to modify it directly, with `reader.offset=X`. That doesn't seem in keeping with the other state, which is all manipulated through functions (which suggests it's meant to be private). Maybe add a `setOffset()` method?

When setting `this.offset`, must the caller also reduce `this.length` by the same amount? The use in `canSeek` suggests no, the use in `peek` (it is missing from the clamp test) clamp suggests yes.

`this.offset` is not used in `expect`, `matchAt`, `assert`, or `findLast`, so I wonder if a client who gets an alleged index back from `findLast` and then submits it to `seek` will find themselves in the wrong place. Some unit tests would be nice.

Reply:
Yes, there’s one location that sets `writer.offset` from the outside. The `writer.length` is orthogonal to `writer.offset`. I could take more care to use private fields. I have a bit of an allergy though: they seem to stand in a weird place between the monkey-patch and go-fast extreme of an everything-public API and the actually-encapsulated extreme with closures or weak maps that plays well with proxies. A push in any direction would be a kindness.

All of the mentioned methods do ultimately take into account the offset without mentioning it expressly:

- `assert` defers to
  - `expect` defers to
    - `matchAt` defers to
      - `byteAt`, which accounts for the offset.
- `findLast` defers to `matchAt`
- `seek` defers to
  - `assertCanSeek` defers to
    - `canSeek` accounts for the offset.

Reply:
Privatized the private data of the Buffer* classes and left eponymous setters with bounds checks where sensible.